File: server/cost/report.js
// The read side of cost tracking — what check_spending.js (and, later,
// self-model.js's own dimension work, if ever extended to cover cost) asks
// to answer "how much have I spent," "which model do I use most." Every
// answer is built from real cost_events rows plus whatever real prices/
// balances are on record — nothing here estimates or interpolates.
//
// Three labelled sections, always kept separate, per the owner's own
// explicit requirement (see cost-store.js's header comment):
//   - measured           — real counted usage, always available.
//   - providerReported    — a real balance/usage figure straight from a
//                           provider's own API, only where one exists.
//   - calculated          — measured × a known price, only where a price is
//                           on record; null (never a guess) otherwise.

import { listEventsSince, listBalances } from './cost-store.js';
import { calculate } from './prices.js';

function startOfMonthIso(now = new Date()) {
  return new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), 1)).toISOString();
}

function startOfDayIso(now = new Date()) {
  return new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate())).toISOString();
}

/**
 * The full measured+calculated breakdown since `sinceIso` — one row per
 * (provider, modelId, unitKind) group. `usdTotal` sums only the groups that
 * actually had a known price; `pricelessGroups` lists which groups
 * contributed no dollar figure, so a caller (or the model relaying this to
 * the user) can say plainly "X of this has no known price" instead of
 * silently under-reporting.
 */
export function usageBreakdown(sinceIso) {
  const events = listEventsSince(sinceIso);
  const groups = new Map(); // key: provider|modelId|unitKind

  for (const ev of events) {
    const key = `${ev.provider}|${ev.modelId || ''}|${ev.unitKind}`;
    const g = groups.get(key) || {
      provider: ev.provider,
      modelId: ev.modelId,
      unitKind: ev.unitKind,
      unitsIn: 0,
      unitsOut: 0,
      cachedIn: 0,
      calls: 0,
    };
    g.unitsIn += ev.unitsIn || 0;
    g.unitsOut += ev.unitsOut || 0;
    g.cachedIn += ev.cachedIn || 0;
    g.calls += 1;
    groups.set(key, g);
  }

  const rows = [];
  let usdTotal = 0;
  let usdCurrency = null;
  const pricelessGroups = [];

  for (const g of groups.values()) {
    const calc = calculate({ provider: g.provider, modelId: g.modelId, unitKind: g.unitKind, unitsIn: g.unitsIn, unitsOut: g.unitsOut });
    if (calc) {
      usdTotal += calc.amount;
      usdCurrency = usdCurrency || calc.currency;
      rows.push({ ...g, calculated: { amount: calc.amount, currency: calc.currency, priceSource: calc.priceSource } });
    } else {
      pricelessGroups.push({ provider: g.provider, modelId: g.modelId, unitKind: g.unitKind });
      rows.push({ ...g, calculated: null });
    }
  }

  return {
    since: sinceIso,
    eventCount: events.length,
    measured: { groups: rows },
    calculated: rows.some((r) => r.calculated) ? { amount: usdTotal, currency: usdCurrency || 'USD' } : null,
    pricelessGroups,
  };
}

export function monthToDate(now = new Date()) {
  return usageBreakdown(startOfMonthIso(now));
}

export function today(now = new Date()) {
  return usageBreakdown(startOfDayIso(now));
}

/** Which model/provider has the most recorded turns since `sinceIso` — a real count, never inferred. */
export function mostUsed(sinceIso) {
  const { measured } = usageBreakdown(sinceIso);
  const sorted = [...measured.groups].sort((a, b) => b.calls - a.calls);
  return sorted[0] || null;
}

/** The last known real balance reading for every provider this build polls — see balances.js. Empty until refreshAllBalances() has run at least once. */
export function providerBalances() {
  return listBalances();
}

File: backend/jarvis/cost/report.py
"""The read side: what has actually been spent, built only from real rows.

Three labelled sections that are never merged — see this package's docstring.
`pricelessGroups` is the honest part: a group with no price on record is listed
by name so the answer can say "this much of it has no known price" rather than
quietly under-reporting the total.

`freeGroups` is the other half of that honesty. A group that cost $0 because the
model is genuinely free is reported as free, by name — not left to be inferred
from an amount of 0.00, which a paid model also produces on a quiet enough day.
Free, unpriced and "used but cost nothing this period" are three different facts.
"""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
from typing import Any

from ..jscompat import to_iso_z
from . import store
from .prices import calculate


def _start_of_day(now: datetime) -> str:
    return to_iso_z(datetime(now.year, now.month, now.day, tzinfo=timezone.utc))


def _start_of_month(now: datetime) -> str:
    return to_iso_z(datetime(now.year, now.month, 1, tzinfo=timezone.utc))


def usage_breakdown(since_iso: str) -> dict[str, Any]:
    events = store.list_events_since(since_iso)
    groups: dict[tuple[str, str | None, str], dict[str, Any]] = {}

    for event in events:
        key = (event["provider"], event["modelId"], event["unitKind"])
        group = groups.setdefault(key, {
            "provider": event["provider"], "modelId": event["modelId"],
            "unitKind": event["unitKind"], "unitsIn": 0, "unitsOut": 0,
            "cachedIn": 0, "calls": 0,
        })
        group["unitsIn"] += event["unitsIn"] or 0
        group["unitsOut"] += event["unitsOut"] or 0
        group["cachedIn"] += event["cachedIn"] or 0
        group["calls"] += 1

    rows: list[dict[str, Any]] = []
    priceless: list[dict[str, Any]] = []
    free: list[dict[str, Any]] = []
    total = 0.0
    currency: str | None = None
    priced = False

    for group in groups.values():
        money = calculate(provider=group["provider"], model_id=group["modelId"],
                          unit_kind=group["unitKind"], units_in=group["unitsIn"],
                          units_out=group["unitsOut"])
        named = {"provider": group["provider"], "modelId": group["modelId"],
                 "unitKind": group["unitKind"]}
        if money is None:
            priceless.append(named)
            rows.append({**group, "calculated": None})
            continue
        priced = True
        total += money["amount"]
        currency = currency or money["currency"]
        if money.get("free"):
            free.append(named)
        rows.append({**group, "calculated": money})

    return {
        "since": since_iso,
        "eventCount": len(events),
        "measured": {"groups": rows},
        # `priced` rather than "any row has a non-zero amount": a real $0 total
        # made entirely of free models is a known total, not a missing one.
        "calculated": ({"amount": total, "currency": currency or "USD"}
                       if priced else None),
        "pricelessGroups": priceless,
        "freeGroups": free,
        "providerReported": store.list_balances(),
    }


def _now(now: datetime | None = None) -> datetime:
    return now or datetime.now(timezone.utc)


def today(now: datetime | None = None) -> dict[str, Any]:
    return usage_breakdown(_start_of_day(_now(now)))


def month_to_date(now: datetime | None = None) -> dict[str, Any]:
    return usage_breakdown(_start_of_month(_now(now)))


def last_days(days: int, now: datetime | None = None) -> dict[str, Any]:
    since = _now(now) - timedelta(days=max(1, days))
    return usage_breakdown(to_iso_z(since))


def most_used(since_iso: str) -> dict[str, Any] | None:
    """Which model was used most — a real count of recorded calls, never inferred
    from anything else."""
    groups = usage_breakdown(since_iso)["measured"]["groups"]
    return max(groups, key=lambda g: g["calls"], default=None)
